### PR TITLE
CEXT-1699 Draft version

### PR DESCRIPTION
**Pull request by MayurBagwe**
## Purpose of this pull request

https://jira.corp.adobe.com/browse/CEXT-1699

## Affected pages

https://developer.adobe.com/commerce/extensibility/events/

## Links to Magento Open Source code
https://github.com/magento-commerce/commerce-eventing/tree/CEXT-1699-NovTest

### CODE REVIEW

File: src/pages/events/convert-existing-fields.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  }
+}
+```
+## Converter definition
+
+The converter class must implement `FieldConverterInterface`. This interface contains the `convert` method, which accepts the following arguments. It returns a mixed type.
+
+`public function convert(mixed $value, Event $event): mixed`
+
+In the following example, the `TestConverterName` converter class updates the value of the `name` field in the `eventData` array.
```
Comment (keharper) [inline src/pages/events/convert-existing-fields.md:77]:
```suggestion
In the following example, the `TestConverterName` converter class updates the value of the `name` field in the `eventData` array to `Test Name`. In the sample payload above, this value was `Men`.
```

File: src/pages/events/convert-existing-fields.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+In the provided code excerpt, the initial value of the `name` field was `Men` and after applying the converter, the updated value is now `Test Name`
```
Comment (keharper) [inline src/pages/events/convert-existing-fields.md:79]: Delete this line. Make sure there's a blank line between the previous paragraph and the code sample.

File: src/pages/events/convert-existing-fields.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+     *
+     * @param mixed $value
+     * @return mixed
+     */
+    public function convert(mixed $value, Event $event): mixed
+    {
+        return 'Test Name';
+    }
+}
+```
```
Comment (keharper) [inline src/pages/events/convert-existing-fields.md:105]: Blank line at the end of the file.

**Review by keharper — CHANGES_REQUESTED** [review]
I need you to do the following outside of this file. 

* Add the `converter` option to the `events:subscribe` command.
* Add this new topic to the TOC. The TOC is in `src/data/navigation/sections/events.js`. Add the entry after line 29.

File: src/pages/events/convert-existing-fields.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+---
+
+# Convert subscribed field value
+
+Your custom module might generate data that would be useful to insert into an existing Commerce event. Converters allow you to enrich the data contained in an event before it is transmitted to the eventing service. 
+
+<InlineAlert variant="info" slots="text"/>
+
+You must configure a module's `io_events.xml` or root `app/etc/io_events.xml` file to update the required fields. You can also declare them in the system `config.php` file or add them using the CLI while subsrcibing to an event.
+
```
[the remark this answers is not in the document]
Comment (MayurBagwe) [reply src/pages/events/convert-existing-fields.md:20]: Updated to - 
`
The event payload might include field values that are not easily interpretable for third-party implementations. Building the implementation solely based on identifiers can be challenging. To address this issue, a converter can be implemented, allowing customization for each field in the payload.`

File: src/pages/events/convert-existing-fields.md [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+`bin/magento events:subscribe observer.catalog_category_save_after --fields="name" --fields='{"name":"store_id", "converter": "Magento\AdobeCommerceEventsClient\Event\TestConverterStoreid"}'`
+
+## Configure the `io_events.xml` file
+
+The `converter` attribute defines the converter class that updates the event data field value for the specified event. Only one converter class can be used per field
+
+Attribute | Required | Description
+--- | --- | ---
+`converter` | No | The fully-qualified class name.
+
```
[the remark this answers is not in the document]
Comment (MayurBagwe) [reply src/pages/events/convert-existing-fields.md:39]: Deleted

File: src/pages/events/convert-existing-fields.md [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+```json
+{
+    "value": {
+       "name": "Test Name",
+       "store_id": "2",
+       "is_active": "1"
+       
+  }
+}
+```
```
[the remark this answers is not in the document]
Comment (MayurBagwe) [reply src/pages/events/convert-existing-fields.md:76]: Added

File: src/pages/events/convert-existing-fields.md [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+title: Convert exisiting fields for the event payload
+description: Learn how to convert a field value based on a condition
+keywords:
+  - Events
+  - Extensibility
+---
+
+# Convert subscribed field value
+
+Your custom module might generate data that would be useful to insert into an existing Commerce event. Converters allow you to enrich the data contained in an event before it is transmitted to the eventing service.
```
[the remark this answers is not in the document]
Comment (MayurBagwe) [reply src/pages/events/convert-existing-fields.md]: Updated - `The event payload might include field values that are not easily interpretable for third-party implementations. Building the implementation solely based on identifiers can be challenging. To address this issue, a converter can be implemented, allowing customization for each field in the payload.`

File: src/pages/events/convert-existing-fields.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (MayurBagwe) [reply src/pages/events/convert-existing-fields.md:20]: Updated

File: src/pages/events/convert-existing-fields.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+<InlineAlert variant="info" slots="text"/>
+
+You must configure a module's `io_events.xml` or root `app/etc/io_events.xml` file to update the required fields. You can also declare them in the system `config.php` file or add them using the CLI while subsrcibing to an event.
+
+## Command line
+
+The `bin/magento events:subscribe <event_code> --force --fields=<name1> --fields='{"<name>":"<name2>", "converter":"<path\to\converterclass>"}'  --fields='{"<name>":"<name3>", "converter":"<path\to\converterclass>"}'` command creates and registers custom and native Commerce events.
+
+## CLI Example
```
[the remark this answers is not in the document]
Comment (MayurBagwe) [reply src/pages/events/convert-existing-fields.md]: Deleted

File: src/pages/events/convert-existing-fields.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+`bin/magento events:subscribe observer.catalog_category_save_after --fields="name" --fields='{"name":"store_id", "converter": "Magento\AdobeCommerceEventsClient\Event\TestConverterStoreid"}'`
```
[the remark this answers is not in the document]
Comment (MayurBagwe) [reply src/pages/events/convert-existing-fields.md]: Updated to use `store_id` field

File: src/pages/events/convert-existing-fields.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+The event payload will be similar to the following without converter:
+
+```json
+{
+    "value": {
+       "name": "Men",
+       "store_id": "2",
+       "is_active": "1"
+
```
[the remark this answers is not in the document]
Comment (MayurBagwe) [reply src/pages/events/convert-existing-fields.md]: Deleted

File: src/pages/events/convert-existing-fields.md (hunk 2 above)
Comment (MayurBagwe) [reply src/pages/events/convert-existing-fields.md]: Updated

File: src/pages/events/convert-existing-fields.md (hunk 3 above)
Comment (MayurBagwe) [reply src/pages/events/convert-existing-fields.md:105]: Added

File: src/pages/events/convert-field-values.md [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+  - Extensibility
+---
+
+# Convert payload field values
+
+The payload of an event often includes field values that are not easily interpretable for third-party implementations. For example, a value might be stored in the Commerce database as an integer, but the external system stores the same information as a string. Alternatively, instead of transforming data types, you might want to change a Commerce-supplied string to a string defined in the external system. To address this issue, you can implement a converter, enabling custom values for any field in the payload.
+
+## Converter definition
+
+Your converter class must implement `FieldConverterInterface`. This interface contains the `convert` method, which accepts `mixed` and `Event` arguments following arguments. The method returns a `mixed` data type. Individual converter classes must be created for each field when aiming to convert multiple fields within a payload. Conversely, the same class can be reused to replace a specific field across multiple events.
```
Comment (MayurBagwe) [inline src/pages/events/convert-field-values.md:15]: Add text that describes advanced usage. Do you have to create a separate class for each field when you want to convert multiple fields within a payload? If you want to replace field X across multiple events, can you reuse the same class? Maybe this is elementary programming, but I'd like to state it.
 
 `Individual converter classes must be created for each field when aiming to convert multiple fields within a payload. Conversely, the same class can be reused to replace a specific field across multiple events.`

File: src/pages/events/convert-field-values.md [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+### Command line
+
+The [`bin/magento events:subscribe --fields` command](commands.md#subscribe-to-an-event) defines the fields and converters to include in the payload of a subscribed event. The example command adds the `visibility` field and provides the path to the converter class. You can specify multiple fields in the same request.
+
+```bash
+bin/magento events:subscribe observer.catalog_product_save_after --fields="store_id" --fields='{"name":"visibility", "converter": "Magento\AdobeCommerceEventsClient\Event\TestConverterVisibility"}'`
+```
+### `config.php` file
+
+```php
```
Comment (MayurBagwe) [inline src/pages/events/convert-field-values.md:110]: Example of config.php file it the event is subscribed using the bash command on [line 106](https://github.com/AdobeDocs/commerce-extensibility/pull/131/files#diff-36693f389686ca3991236ee3437b72b755c9770cb875d50f78f211d1c76b4ec6R106)

File: src/pages/events/convert-field-values.md [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+                'store_id',
+                [
+                    'name' => 'visibility',
+                    'converter' => 'Magento\\AdobeCommerceEventsClient\\Event\\TestConverterVisibility'
+                ]
+            ],
+            'enabled' => 1
+        ]
+ ]       
+```
```
Comment (keharper) [inline src/pages/events/convert-field-values.md:128]: Add a blank line above the head3

File: src/pages/events/convert-field-values.md [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+You must configure a module's `io_events.xml` or root `app/etc/io_events.xml` file to update the required fields. You can also declare them in the system `config.php` file or add them using the CLI while subscribing to an event.
+
+### Command line
+
+The [`bin/magento events:subscribe --fields` command](commands.md#subscribe-to-an-event) defines the fields and converters to include in the payload of a subscribed event. The example command adds the `visibility` field and provides the path to the converter class. You can specify multiple fields in the same request.
+
+```bash
+bin/magento events:subscribe observer.catalog_product_save_after --fields="store_id" --fields='{"name":"visibility", "converter": "Magento\AdobeCommerceEventsClient\Event\TestConverterVisibility"}'`
+```
+### `config.php` file
```
Comment (keharper) [inline src/pages/events/convert-field-values.md:113]: Add a blank line above the head3

Add the following line below the head3.

The following example `config.php` is the equivalent of the example `bin/magento events:subscribe` command in the **Command line** example above.

Also make sure there's one blank line above and below the new sentence.

File: src/pages/events/convert-field-values.md (hunk 11 above)
Comment (keharper) [inline src/pages/events/convert-field-values.md:15]:
```suggestion
Your converter class must implement `FieldConverterInterface`. This interface contains the `convert` method, which accepts `mixed` and `Event` arguments. The method returns a `mixed` data type. You must create individual converter classes for each field when converting multiple fields within a payload. You can reuse a class to replace a specific field across multiple events.
```

File: src/pages/events/convert-field-values.md (hunk 13 above)
Comment (MayurBagwe) [reply src/pages/events/convert-field-values.md:131]: Added

File: src/pages/events/convert-field-values.md (hunk 14 above)
Comment (MayurBagwe) [reply src/pages/events/convert-field-values.md:114]: Added

**Review by keharper — APPROVED** [review]

File: src/pages/events/commands.md [hunk 15]
```diff
@@ -203,6 +203,8 @@ If you are implementing eventing in a performance testing environment, run the `
 
 `--rules=<field-name>|<operator>|<value>` Defines a rule that will be applied to the subscribed event. You can apply multiple rules to an event, but each rule must be defined separately. A rule definition must specify the field to be evaluated, an operator, and the value to be evaluated, in that order. The field name in a rule definition does not have to match a field specified with the `--fields` option.
 
+`--fields='{"<name>":"<field-name>", "converter":"<path\to\converterclass>"}'` Applies the converter class to the given field.
```
Comment (jhadobe) [inline src/pages/events/commands.md:206]: You might reorder the options (alphabetically with required first?).

File: src/pages/events/convert-field-values.md [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+     *
+     * @param mixed $value
+     * @param Event $event
+     * @return mixed
+     */
+    public function convert(mixed $value, Event $event);
+}
+```
+
+As an example, the `observer.catalog_product_save_after` event contains a top-level `visibility` field, which must contain an integer value. You want to convert these values to strings that match values on the external system. The following table describes these values.
```
Comment (jhadobe) [inline src/pages/events/convert-field-values.md:31]:
```suggestion
As an example, the `observer.catalog_product_save_after` event contains a top-level `visibility` field, which must contain an integer value. Convert these values to strings that match values on the external system. The following table describes these values.
```

File: src/pages/events/convert-field-values.md [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+            "visibility":"CATALOG_AND_SEARCH",
+        }
+      }
+   }
+}
+```
+
+## Register the converter
+
+You must configure a module's `io_events.xml` or root `app/etc/io_events.xml` file to update the required fields. You can also declare them in the system `config.php` file or add them using the CLI while subscribing to an event.
```
Comment (jhadobe) [inline src/pages/events/convert-field-values.md:104]:
```suggestion
You must configure a module's `io_events.xml` or root `app/etc/io_events.xml` file to update the required fields. You can also declare them in the system `config.php` file or add them when using the CLI to subscribe to an event.
```

**Review by jhadobe — APPROVED** [review]

**Pull request merged**